### PR TITLE
[MM-45116] Feature: playbooks insights

### PR DESCRIPTION
**Pull request by plant99**
#### Summary

Add two new endpoints used by Insights page in mm-webapp.
```
/playbooks/api/v0/playbooks/insights/teams/4f3rbqs5qibn5erzfazxgmssmh?time_range=28_day&page=0&per_page=200
/playbooks/api/v0/playbooks/insights/user/me?time_range=28_day&page=0&per_page=200&team_id=4f3rbqs5qibn5erzfazxgmssmh
```
Following is a sample response
```
{
  "has_next": false,
  "items": [
    {
      "playbook_id": "6nrnwj5hfty8787mos9h6dtx6y",
      "num_runs": 3,
      "title": "sth",
      "last_run_at": 1658827931570
    },
    {
      "playbook_id": "3ii9fwq47pdtupqdd6ajo3axuc",
      "num_runs": 1,
      "title": "Test playbook",
      "last_run_at": 1658493052384
    }
  ]
}
```
#### Ticket Link

https://mattermost.atlassian.net/browse/MM-45116
#### Checklist

- [ ] Telemetry updated
- [ ] Gated by experimental feature flag
- [x] Unit tests updated

### CODE REVIEW

File: server/api/playbooks.go [hunk 1]
```diff
@@ -609,3 +615,102 @@ func (h *PlaybookHandler) validateMetrics(pb app.Playbook) error {
 	}
 	return nil
 }
+
+func (h *PlaybookHandler) getTopPlaybooksForUser(w http.ResponseWriter, r *http.Request) {
+	userID := r.Header.Get("Mattermost-User-ID")
+	params := r.URL.Query()
+	timeRange := params.Get("time_range")
+	teamID := params.Get("team_id")
+	if teamID != "" && !h.PermissionsCheck(w, h.permissions.PlaybookList(userID, teamID)) {
```
Comment (crspeller) [inline server/api/playbooks.go:624]: What if `teamID == ""`? What is the expected behavior? Should we return an error?

**Review by crspeller — CHANGES_REQUESTED** [review]

File: server/sqlstore/playbook.go [hunk 2]
```diff
@@ -1096,3 +1097,95 @@ func toPlaybook(rawPlaybook sqlPlaybook) (app.Playbook, error) {
...
+
+// GetTopBoardInsightsListWithPagination adds a rank to each item in the given list of PlaybooksInsight and checks if there is
+// another page that can be fetched based on the given limit and offset. The given list of PlaybooksInsight is assumed to be
+// sorted by Runs(score). Returns a PlaybooksInsightsList.
+func GetTopPlaybooksInsightsListWithPagination(boards []*app.PlaybookInsight, limit int) *app.PlaybooksInsightsList {
+	// Add pagination support
+	var hasNext bool
+	if (limit != 0) && (len(boards) == limit+1) {
+		hasNext = true
+		boards = boards[:len(boards)-1]
```
[the remark this answers is not in the document]
Comment (plant99) [reply server/sqlstore/playbook.go]: Thanks for pointing this out, [here](https://github.com/mattermost/mattermost-plugin-playbooks/pull/1366/files#diff-54c87c177e0c1f4e0ab4c2c0c903e13fcdabf27aa1e12afae9801c66ae1b95f5R1135) we query for an extra row so we can determine if `Page.HasNext` is true or false.

File: server/sqlstore/playbook.go [hunk 3]
```diff
@@ -1096,3 +1097,95 @@ func toPlaybook(rawPlaybook sqlPlaybook) (app.Playbook, error) {
...
+			"p.Title",
+			"COUNT(i.ID) AS NumRuns",
+			"COALESCE(MAX(i.CreateAt), 0) AS LastRunAt",
+		).
+		From("IR_Playbook as p").
+		LeftJoin("IR_Incident AS i ON p.ID = i.PlaybookID").
+		Where(sq.And{
+			sq.GtOrEq{"i.CreateAt": opts.StartUnixMilli},
+			sq.Or{
+				sq.Eq{"p.ID": accessiblePlaybooks},
```
[the remark this answers is not in the document]
Comment (plant99) [reply server/sqlstore/playbook.go]: Please let me know if the query seems wrong to you with the latest commit.

File: server/api/playbooks.go (hunk 1 above)
Comment (plant99) [reply server/api/playbooks.go]: Added checks for teamID = ""

**Comment by plant99** [comment]
Thanks so much @crspeller for the review! I've updated the code as per your suggestions.

**Review by mkraft — APPROVED** [review]

File: server/sqlstore/playbook.go [hunk 4]
```diff
@@ -1096,3 +1096,107 @@ func toPlaybook(rawPlaybook sqlPlaybook) (app.Playbook, error) {
...
+// insights - store manager functions
+
+func (p *playbookStore) GetTopPlaybooksForTeam(teamID, userID string, opts *model.InsightsOpts) (*app.PlaybooksInsightsList, error) {
+
+	permissionsAndFilter := sq.Expr(`(
+		EXISTS(SELECT 1
+				FROM IR_PlaybookMember as pm
+				WHERE pm.PlaybookID = p.ID
+				AND pm.MemberID = ?)
+		OR NOT EXISTS(SELECT 1
```
Comment (crspeller) [inline server/sqlstore/playbook.go:1109]: This `OR NOT EXISTS` part is unnecessary. No members used to mean public but that is no longer the case.

File: server/sqlstore/playbook.go [hunk 5]
```diff
@@ -1096,3 +1096,107 @@ func toPlaybook(rawPlaybook sqlPlaybook) (app.Playbook, error) {
...
+}
+
+func (p *playbookStore) GetTopPlaybooksForUser(teamID, userID string, opts *model.InsightsOpts) (*app.PlaybooksInsightsList, error) {
+
+	permissionsAndFilter := sq.Expr(`(
+		EXISTS(SELECT 1
+				FROM IR_PlaybookMember as pm
+				WHERE pm.PlaybookID = p.ID
+				AND pm.MemberID = ?)
+		OR NOT EXISTS(SELECT 1
```
Comment (crspeller) [inline server/sqlstore/playbook.go:1155]: Same here.

File: server/sqlstore/playbook.go [hunk 6]
```diff
@@ -1096,3 +1096,107 @@ func toPlaybook(rawPlaybook sqlPlaybook) (app.Playbook, error) {
...
+			"COUNT(i.ID) AS NumRuns",
+			"COALESCE(MAX(i.CreateAt), 0) AS LastRunAt",
+		).
+		From("IR_Playbook as p").
+		LeftJoin("IR_Incident AS i ON p.ID = i.PlaybookID").
+		Where(sq.And{
+			sq.GtOrEq{"i.CreateAt": opts.StartUnixMilli},
+			sq.Or{
+				permissionsAndFilter,
+				sq.Eq{"p.Public": true},
```
Comment (crspeller) [inline server/sqlstore/playbook.go:1128]: This query and the one in GetTopPlaybooksForUser are identical except for this one includes all public playbooks. Could we construct a common query and just modify the one where clause?

**Review by crspeller — CHANGES_REQUESTED** [review]

File: server/sqlstore/playbook.go (hunk 6 above)
Comment (plant99) [reply server/sqlstore/playbook.go]: Nice suggestion, thank you! For now, I've generalised just the query builder. But I see scope of them being a common function with params changing in app layer. Not sure if it's a good pattern. I'd love to know your thoughts on that.

**Review by BenCookie95 — APPROVED** [review]

**Review by crspeller — APPROVED** [review]

**Comment by lieut-data** [comment]
@Phrynobatrachus, can you help take a look at this PR for the Insights feature?

**Comment by Phrynobatrachus** [comment]
Is the feature supposed to be fully working alongside https://github.com/mattermost/mattermost-webapp/pull/10919/files or is there still additional plumbing left to do?

**Review by Phrynobatrachus — APPROVED** [review]
Endpoints themselves seem fine so approving this, if my question above is an actual issue I expect it's on the webapp side.

**Comment by BenCookie95** [comment]
> Is the feature supposed to be fully working alongside https://github.com/mattermost/mattermost-webapp/pull/10919/files or is there still additional plumbing left to do? As it is right now, the table only shows the loading state, and no actual data.
> 
> Screenshot from 2022-08-19 13-13-09

I'll take a look and see what's up!

**Comment by BenCookie95** [comment]
Oh I see what happened, we merged the client stuff into upstream and not Shivashis' branch.

@plant99 Can you merge upstream/MM-45116_top_playbooks into this branch?

**Comment by plant99** [comment]
Hi @BenCookie95 , @Phrynobatrachus  I've merged changes from that branch into this one. Should work now.

**Pull request merged**